src/psyclone/psyir/symbols/datatypes.py

```diff
@@ -346,6 +362,58 @@ def __str__(self):
             self._datatype, ", ".join(dims)))
 
 
+class StructureType(DataType):
+    '''
+    Describes a 'structure' or 'derived' datatype that is itself composed
+    of a list of other datatypes. Those datatypes are stored as an
+    OrderedDict of namedtuples.
+
+    '''
+    # Each member of a StructureType is represented by a ComponentType
+    # (named tuple).
+    ComponentType = namedtuple("ComponentType", ["name", "datatype",
```

Comment:
This implementation turned out to be leaner that I imagined. But I do see the similarities with the symbol table.

I think the current implementation is fine but I would add a note here or in the developer's guide about this decision, saying that if the duplicated features/functionality grows we may re-evaluate this decision. What do you think?

Reply:
I've added a note in the class docstring.